### PR TITLE
Improve nats reconciler when subscription filters are modified

### PR DESCRIPTION
**Pull request by mfaizanse**
**Description**

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/pkg/handlers/nats.go [hunk 1]
```diff
@@ -261,7 +256,7 @@ func (n *Nats) deleteSubFromNats(natsSub *nats.Subscription, subKey string, log
 		}
 	}
 	delete(n.subscriptions, subKey)
-	log.Debugw("unsubscribe succeeded", "subscriptionKey", subKey)
+	log.Debugw("unsubscribe from nats succeeded", "subscriptionKey", subKey)
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats.go:259]: nit: NATS

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 2]
```diff
@@ -617,6 +614,287 @@ func TestNatsSubAfterSync_FiltersChange(t *testing.T) {
...
+	// The sink should receive the event for second subject
+	err = subscriber.CheckEvent(expectedDataInStore, subscriberCheckURL)
+	if err != nil {
+		t.Fatalf("check event failed for second subject: %v", err)
+	}
+}
+
+// TestNatsSubAfterSync_FilterRemoved tests the SyncSubscription method
+// when a filter is removed from subscription
+func TestNatsSubAfterSync_FilterRemoved(t *testing.T) {
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats_test.go:739]: can we have a test with multiple subscriptions? 
1 sub will not be modified
1 sub will be modified
both should work as expected before and after modification

if this test already exists, please ignore, i could not find it

Comment (mfaizanse) [reply components/eventing-controller/pkg/handlers/nats_test.go:757]: done

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 3]
```diff
@@ -895,6 +895,172 @@ func TestNatsSubAfterSync_FilterRemoved(t *testing.T) {
...
+	g := NewWithT(t)
+
+	// // ######  Setup test assets ######
+	// setup logger
+	defaultLogger, err := logger.New(string(kymalogger.JSON), string(kymalogger.INFO))
+	if err != nil {
+		t.Fatalf("initialize logger failed: %v", err)
+	}
+
+	// create subscribers servers for testing
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats_test.go:945]:
```suggestion
```

File: components/eventing-controller/pkg/handlers/nats_test.go [hunk 4]
```diff
@@ -617,6 +614,453 @@ func TestNatsSubAfterSync_FiltersChange(t *testing.T) {
...
+	natsPort := nextPort.get()
+	subscriberPort := nextPort.get()
+	subscriberReceiveURL := fmt.Sprintf("http://127.0.0.1:%d/store", subscriberPort)
+	subscriberCheckURL := fmt.Sprintf("http://127.0.0.1:%d/check", subscriberPort)
+
+	// Create a new subscriber
+	subscriber := eventingtesting.NewSubscriber(fmt.Sprintf(":%d", subscriberPort))
+	subscriber.Start()
+	defer subscriber.Shutdown() // defer the shutdown of subscriber
+	defer subscriber.Shutdown()
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats_test.go:956]:
```suggestion
```
it should be dead already

**Review by k15r — APPROVED** [review]

**Pull request merged**